contracts/0.8.9/WithdrawalQueue.sol

```diff
@@ -0,0 +1,639 @@
+
+        return request.recipient;
+    }
+
+    /**
+     * @notice calculates the params to fulfill the next batch of requests in queue
+     * @param _lastIdToFinalize last id in the queue to finalize upon
+     * @param _totalPooledEther share price component to finalize requests
+     * @param _totalShares share price component to finalize requests
+     *
+     * @return etherToLock amount of eth required to finalize the batch
+     * @return sharesToBurn amount of shares that should be burned on finalization
+     */
+    function calculateFinalizationParams(
+        uint256 _lastIdToFinalize,
+        uint256 _totalPooledEther,
+        uint256 _totalShares
+    ) external view returns (uint256 etherToLock, uint256 sharesToBurn) {
+        return _calculateDiscountedBatch(
+            finalizedRequestsCounter,
+            _lastIdToFinalize,
+            _toUint128(_totalPooledEther),
+            _toUint128(_totalShares)
+        );
+    }
+
+    function findPriceHint(uint256 _requestId) public view returns (uint256 hint) {
+        if (_requestId >= finalizedRequestsCounter) revert PriceNotFound();
+
+        for (uint256 i = finalizationPrices.length; i > 0; i--) {
+            if (_isPriceHintValid(_requestId, i - 1)) {
+                return i - 1;
+            }
+        }
+        assert(false);
+    }
+
+    function restake(uint256 _amount) external onlyOwner {
+        if (lockedEtherAmount + _amount > address(this).balance) revert NotEnoughEther();
+
+        IRestakingSink(OWNER).receiveRestake{value: _amount}();
+    }
+
+    function _calculateDiscountedBatch(
+        uint256 firstId,
+        uint256 lastId,
+        uint128 _totalPooledEther,
+        uint128 _totalShares
+    ) internal view returns (uint128 eth, uint128 shares) {
+        eth = queue[lastId].cumulativeEther;
+        shares = queue[lastId].cumulativeShares;
+
+        if (firstId > 0) {
+            eth -= queue[firstId - 1].cumulativeEther;
+            shares -= queue[firstId - 1].cumulativeShares;
+        }
+
+        eth = _min(eth, (shares * _totalPooledEther) / _totalShares);
+    }
+
+    function _isPriceHintValid(uint256 _requestId, uint256 hint) internal view returns (bool isInRange) {
+        uint256 hintLastId = finalizationPrices[hint].index;
+
+        isInRange = _requestId <= hintLastId;
+        if (hint > 0) {
+            uint256 previousId = finalizationPrices[hint - 1].index;
+
+            isInRange = isInRange && previousId < _requestId;
+        }
+    }
+
+    function _updatePriceHistory(uint128 _totalPooledEther, uint128 _totalShares, uint256 index) internal {
+        if (finalizationPrices.length == 0) {
+            finalizationPrices.push(Price(_totalPooledEther, _totalShares, index));
+        } else {
+            Price storage lastPrice = finalizationPrices[finalizationPrices.length - 1];
+
+            if (_totalPooledEther / _totalShares == lastPrice.totalPooledEther / lastPrice.totalShares) {
```

Comment:
won't work in the case of `totalPooledEther < totalShares`
Requires chainlink-like price format (e.g., 10^27 when `totalPooledEther == totalShares`)